File: android/app/src/main/java/com/emwaver/emwaverandroidapp/DeviceBufferSession.java
/*
 * EMWaver
 * Copyright (c) 2026 Luís Marnoto
 * SPDX-License-Identifier: Apache-2.0
 */

package com.emwaver.emwaverandroidapp;

import java.io.ByteArrayOutputStream;
import java.util.Arrays;

final class DeviceBufferSession implements TransportDeviceSession {
    private static final int PACKET_SIZE_BYTES = 18;
    private static final int EMW_OP_SAMPLE = 0x60;
    private static final int EMW_SAMPLE_START = 0x00;
    private static final int EMW_SAMPLE_STOP = 0x01;

    private byte[] rxBytes = new byte[0];
    private long[] rxTsMs = new long[0];
    private long rxCounter = 0;
    private byte[] txBytes = new byte[0];
    private long[] txTsMs = new long[0];
    private boolean samplerStreamingActive = false;
    private final ByteArrayOutputStream sysexBuf = new ByteArrayOutputStream(64);
    private boolean inSysex = false;
    private final String deviceId;

    DeviceBufferSession() {
        this("active");
    }

    DeviceBufferSession(String deviceId) {
        this.deviceId = deviceId == null || deviceId.trim().isEmpty() ? "active" : deviceId.trim();
    }

    @Override
    public String deviceId() {
        return deviceId;
    }

    @Override
    public synchronized void clearAll() {
        rxBytes = new byte[0];
        rxTsMs = new long[0];
        rxCounter = 0;
        txBytes = new byte[0];
        txTsMs = new long[0];
        samplerStreamingActive = false;
        sysexBuf.reset();
        inSysex = false;
    }

    @Override
    public synchronized int getBufferLength() {
        return rxBytes.length;
    }

    @Override
    public synchronized void loadBuffer(byte[] data) {
        rxBytes = data != null ? Arrays.copyOf(data, data.length) : new byte[0];
        rxCounter = 0;
        int packets = rxBytes.length / PACKET_SIZE_BYTES;
        rxTsMs = new long[packets];
    }

    @Override
    public synchronized byte[] getBuffer() {
        return Arrays.copyOf(rxBytes, rxBytes.length);
    }

    @Override
    public synchronized void storeBulkPkt(byte[] data, long tsMs) {
        if (data == null || data.length == 0) return;

        int prevPackets = rxBytes.length / PACKET_SIZE_BYTES;
        rxBytes = appendBytes(rxBytes, data);
        int newPackets = rxBytes.length / PACKET_SIZE_BYTES;
        int delta = Math.max(0, newPackets - prevPackets);
        if (delta > 0) {
            rxTsMs = appendRepeated(rxTsMs, tsMs, delta);
        }
    }

    @Override
    public synchronized void appendTxBytes(byte[] data, long tsMs) {
        if (data == null || data.length == 0) return;

        for (int offset = 0; offset < data.length; offset += PACKET_SIZE_BYTES) {
            int take = Math.min(PACKET_SIZE_BYTES, data.length - offset);
            byte[] pkt = new byte[PACKET_SIZE_BYTES];
            System.arraycopy(data, offset, pkt, 0, take);
            txBytes = appendBytes(txBytes, pkt);
            txTsMs = appendRepeated(txTsMs, tsMs, 1);
        }
    }

    @Override
    public synchronized long getTxPacketCount() {
        return txTsMs.length;
    }

    @Override
    public synchronized byte[] getTxBuffer() {
        return Arrays.copyOf(txBytes, txBytes.length);
    }

    @Override
    public synchronized void updateSamplerStreamingState(byte[] lane) {
        if (lane == null || lane.length < 2) {
            return;
        }
        int opcode = lane[0] & 0xFF;
        if (opcode != EMW_OP_SAMPLE) {
            return;
        }
        int sub = lane[1] & 0xFF;
        if (sub == EMW_SAMPLE_START) {
            samplerStreamingActive = true;
        } else if (sub == EMW_SAMPLE_STOP) {
            samplerStreamingActive = false;
        }
    }

    @Override
    public synchronized boolean shouldStoreStreamLane(byte[] streamLane) {
        return !isLaneEmpty(streamLane) || samplerStreamingActive;
    }

    @Override
    public synchronized void resetSamplerStreaming() {
        samplerStreamingActive = false;
    }

    @Override
    public synchronized void feedSysexBytes(byte[] data, int offset, int count, long tsMs) {
        if (data == null || count <= 0) {
            return;
        }

        int end = Math.min(data.length, offset + count);
        for (int i = Math.max(0, offset); i < end; i++) {
            byte b = data[i];
            if (b == (byte) 0xF0) {
                sysexBuf.reset();
                inSysex = true;
            }
            if (!inSysex) {
                continue;
            }
            sysexBuf.write(b);
            if (sysexBuf.size() > 128) {
                sysexBuf.reset();
                inSysex = false;
                continue;
            }
            if (b == (byte) 0xF7) {
                inSysex = false;
                byte[] sysex = sysexBuf.toByteArray();
                sysexBuf.reset();

                byte[] frame = UsbMidiSysex.decodeSysexToFrame(sysex);
                if (frame == null || frame.length != UsbMidiSysex.FRAME_SIZE) {
                    continue;
                }

                byte[] cmdLane = Arrays.copyOfRange(frame, 0, UsbMidiSysex.LANE_SIZE);
                byte[] streamLane = Arrays.copyOfRange(frame, UsbMidiSysex.LANE_SIZE, UsbMidiSysex.FRAME_SIZE);

                if (!isLaneEmpty(cmdLane)) {
                    storeBulkPkt(cmdLane, tsMs);
                }
                if (shouldStoreStreamLane(streamLane)) {
                    storeBulkPkt(streamLane, tsMs);
                }
            }
        }
    }

    @Override
    public synchronized void prepareCommandResponseWait() {
        rxCounter = rxBytes.length / (long) PACKET_SIZE_BYTES;
    }

    @Override
    public byte[] awaitCommandResponse(int timeoutMs) {
        long startTime = System.currentTimeMillis();
        int safeTimeout = Math.max(1, timeoutMs);
        while (System.currentTimeMillis() - startTime < safeTimeout) {
            byte[] pkt = nextRxPacketData();
            if (pkt != null && pkt.length >= PACKET_SIZE_BYTES) {
                int status = pkt[0] & 0xFF;
                if (status >= 0x80) {
                    return Arrays.copyOf(pkt, PACKET_SIZE_BYTES);
                }
            }
            try {
                Thread.sleep(5);
            } catch (InterruptedException ignored) {
                Thread.currentThread().interrupt();
                return null;
            }
        }
        return null;
    }

    @Override
    public synchronized long getRxPacketCount() {
        return rxBytes.length / (long) PACKET_SIZE_BYTES;
    }

    @Override
    public synchronized void setRxCounter(long value) {
        long packets = rxBytes.length / (long) PACKET_SIZE_BYTES;
        long desired = Math.max(0, value);
        rxCounter = Math.min(desired, packets);
    }

    @Override
    public synchronized Object[] nextRxPacket() {
        byte[] pkt = nextRxPacketDataLocked();
        if (pkt == null) return null;
        long ts = rxCounter > 0 && rxCounter - 1 < rxTsMs.length ? rxTsMs[(int) rxCounter - 1] : 0;
        return new Object[] { pkt, ts };
    }

    private synchronized byte[] nextRxPacketData() {
        return nextRxPacketDataLocked();
    }

    private byte[] nextRxPacketDataLocked() {
        long packets = rxBytes.length / (long) PACKET_SIZE_BYTES;
        if (rxCounter >= packets) return null;

        int startByte = (int) (rxCounter * PACKET_SIZE_BYTES);
        if (startByte + PACKET_SIZE_BYTES > rxBytes.length) return null;

        byte[] pkt = Arrays.copyOfRange(rxBytes, startByte, startByte + PACKET_SIZE_BYTES);
        rxCounter += 1;
        return pkt;
    }

    @Override
    public synchronized Object[] takeRxState() {
        return new Object[] { Arrays.copyOf(rxBytes, rxBytes.length), Arrays.copyOf(rxTsMs, rxTsMs.length), rxCounter };
    }

    @Override
    public synchronized void restoreRxState(byte[] rxBytesIn, long[] rxTsMsIn, long rxCounterIn) {
        rxBytes = rxBytesIn != null ? Arrays.copyOf(rxBytesIn, rxBytesIn.length) : new byte[0];
        rxTsMs = rxTsMsIn != null ? Arrays.copyOf(rxTsMsIn, rxTsMsIn.length) : new long[0];

        int packets = rxBytes.length / PACKET_SIZE_BYTES;
        if (rxTsMs.length != packets) {
            rxTsMs = Arrays.copyOf(rxTsMs, packets);
        }

        long desired = Math.max(0, rxCounterIn);
        rxCounter = Math.min(desired, (long) packets);
    }

    @Override
    public synchronized Object[] compressDataBits(int rangeStart, int rangeEnd, int numberBins) {
        int totalBits = rxBytes.length * 8;
        if (rxBytes.length == 0 || rangeStart >= rangeEnd || rangeStart >= totalBits || numberBins <= 0) {
            return new Object[] { new float[0], new float[0] };
        }

        int end = Math.min(rangeEnd, totalBits);
        int start = Math.min(rangeStart, end);
        int span = end - start;
        if (span <= 0) {
            return new Object[] { new float[0], new float[0] };
        }

        if (span <= numberBins * 2) {
            float[] t = new float[span];
            float[] v = new float[span];
            for (int i = 0; i < span; i++) {
                int bitIndex = start + i;
                t[i] = bitIndex;
                v[i] = bitAt(rxBytes, bitIndex) == 1 ? 255f : 0f;
            }
            return new Object[] { t, v };
        }

        float[] tTmp = new float[numberBins * 2];
        float[] vTmp = new float[numberBins * 2];
        int outCount = 0;

        double binWidth = (double) span / (double) numberBins;
        for (int bin = 0; bin < numberBins; bin++) {
            int binStart = (int) Math.floor((double) start + (double) bin * binWidth);
            int binEnd = (int) Math.floor((double) binStart + binWidth);
            if (binEnd > end) binEnd = end;
            if (binEnd <= binStart) continue;

            boolean hasLow = false;
            boolean hasHigh = false;

            int i = binStart;
            while (i < binEnd) {
                int byteIndex = i >> 3;
                if (byteIndex >= rxBytes.length) break;

                if ((i & 7) == 0 && i + 8 <= binEnd) {
                    int byteVal = rxBytes[byteIndex] & 0xFF;
                    if (byteVal == 0) {
                        hasLow = true;
                    } else if (byteVal == 255) {
                        hasHigh = true;
                    } else {
                        hasLow = true;
                        hasHigh = true;
                    }
                    i += 8;
                } else {
                    if (bitAt(rxBytes, i) == 1) hasHigh = true; else hasLow = true;
                    i += 1;
                }

                if (hasLow && hasHigh) break;
            }

            if (hasLow || hasHigh) {
                if (outCount + 2 > tTmp.length) break;
                tTmp[outCount] = binStart;
                vTmp[outCount] = hasLow ? 0f : 255f;
                outCount++;
                tTmp[outCount] = (binEnd - 1);
                vTmp[outCount] = hasHigh ? 255f : 0f;
                outCount++;
            }
        }

        return new Object[] { Arrays.copyOf(tTmp, outCount), Arrays.copyOf(vTmp, outCount) };
    }

    private static int bitAt(byte[] buf, int bitIndex) {
        int byteIndex = bitIndex >> 3;
        if (byteIndex < 0 || byteIndex >= buf.length) return 0;
        int b = buf[byteIndex] & 0xFF;
        int shift = bitIndex & 7;
        return (b >> shift) & 1;
    }

    private static byte[] appendBytes(byte[] a, byte[] b) {
        if (b.length == 0) return a;
        if (a.length == 0) return Arrays.copyOf(b, b.length);
        byte[] out = Arrays.copyOf(a, a.length + b.length);
        System.arraycopy(b, 0, out, a.length, b.length);
        return out;
    }

    private static long[] appendRepeated(long[] a, long value, int count) {
        if (count <= 0) return a;
        long[] out = Arrays.copyOf(a, a.length + count);
        for (int i = 0; i < count; i++) out[a.length + i] = value;
        return out;
    }

    private static boolean isLaneEmpty(byte[] lane) {
        if (lane == null || lane.length == 0) return true;
        for (byte b : lane) {
            if (b != 0) return false;
        }
        return true;
    }
}

File: ios/EMWaver/Managers/DeviceBufferSession.swift
/*
 * EMWaver
 * Copyright (c) 2026 Luís Marnoto
 * SPDX-License-Identifier: Apache-2.0
 */

import Foundation

final class DeviceBufferSession {
    private static let packetSizeBytes = 18
    private let lock = NSLock()

    private var rxBytes: [UInt8] = []
    private var rxTsMs: [UInt64] = []
    private var rxCounter: UInt64 = 0
    private var txBytes: [UInt8] = []
    private var txTsMs: [UInt64] = []
    private var samplerStreamingActive = false
    private var sysexAccumulator = UsbMidiSysexAccumulator()

    func clearAll() {
        lock.lock()
        defer { lock.unlock() }
        rxBytes = []
        rxTsMs = []
        rxCounter = 0
        txBytes = []
        txTsMs = []
        samplerStreamingActive = false
        sysexAccumulator = UsbMidiSysexAccumulator()
    }

    func getRxPacketCount() -> UInt64 {
        lock.lock()
        defer { lock.unlock() }
        return UInt64(rxBytes.count / Self.packetSizeBytes)
    }

    func getTxPacketCount() -> UInt64 {
        lock.lock()
        defer { lock.unlock() }
        return UInt64(txTsMs.count)
    }

    func getRxCounter() -> UInt64 {
        lock.lock()
        defer { lock.unlock() }
        return rxCounter
    }

    func setRxCounter(_ value: UInt64) {
        lock.lock()
        defer { lock.unlock() }
        let packets = UInt64(rxBytes.count / Self.packetSizeBytes)
        rxCounter = min(value, packets)
    }

    func storeBulkPkt(_ data: Data, tsMs: UInt64) {
        if data.isEmpty { return }
        lock.lock()
        defer { lock.unlock() }
        storeBulkPktLocked(data, tsMs: tsMs)
    }

    private func storeBulkPktLocked(_ data: Data, tsMs: UInt64) {
        let prevPackets = rxBytes.count / Self.packetSizeBytes
        rxBytes.append(contentsOf: data)
        let newPackets = rxBytes.count / Self.packetSizeBytes
        let delta = max(0, newPackets - prevPackets)
        if delta > 0 {
            rxTsMs.append(contentsOf: Array(repeating: tsMs, count: delta))
        }
    }

    func appendTxBytes(_ data: Data, tsMs: UInt64) {
        if data.isEmpty { return }
        lock.lock()
        defer { lock.unlock() }

        var offset = 0
        while offset < data.count {
            let take = min(Self.packetSizeBytes, data.count - offset)
            var pkt = [UInt8](repeating: 0, count: Self.packetSizeBytes)
            data.copyBytes(to: &pkt, from: offset..<(offset + take))
            txBytes.append(contentsOf: pkt)
            txTsMs.append(tsMs)
            offset += take
        }
    }

    func outgoingSamplerPolicy(for data: Data) -> SamplerLanePolicy {
        lock.lock()
        defer { lock.unlock() }
        let policy = SamplerLanePolicy.forOutgoingPacket(data, samplerStreamingActive: samplerStreamingActive)
        samplerStreamingActive = policy.nextSamplerStreamingActive
        return policy
    }

    func resetSamplerStreaming() {
        lock.lock()
        defer { lock.unlock() }
        samplerStreamingActive = false
    }

    func feedMidiBytes(_ normalized: Data, tsMs: UInt64) {
        if normalized.isEmpty { return }
        lock.lock()
        defer { lock.unlock() }

        for sysex in sysexAccumulator.feed(normalized) {
            guard let superframe = UsbMidiSysex.decodeSysexToSuperframe(sysex) else {
                continue
            }
            guard superframe.count >= Self.packetSizeBytes * 2 else {
                continue
            }

            let commandLane = superframe.subdata(in: 0..<Self.packetSizeBytes)
            let streamLane = superframe.subdata(in: Self.packetSizeBytes..<(Self.packetSizeBytes * 2))
            let policy = SamplerLanePolicy.forIncomingSuperframe(
                commandLane: commandLane,
                streamLane: streamLane,
                samplerStreamingActive: samplerStreamingActive
            )

            if policy.shouldStoreCommandLane {
                storeBulkPktLocked(commandLane, tsMs: tsMs)
            }
            if policy.shouldStoreStreamLane {
                storeBulkPktLocked(streamLane, tsMs: tsMs)
            }
        }
    }

    func prepareCommandResponseWait() {
        lock.lock()
        defer { lock.unlock() }
        rxCounter = UInt64(rxBytes.count / Self.packetSizeBytes)
    }

    func awaitCommandResponse(timeout: Int, shouldContinue: () -> Bool) -> Data? {
        let startTime = Date().timeIntervalSince1970
        let safeTimeout = Double(max(1, timeout))

        while true {
            if !shouldContinue() { return nil }
            if let pkt = nextRxPacket() {
                return pkt.packet64
            }

            let elapsedMs = (Date().timeIntervalSince1970 - startTime) * 1000
            if elapsedMs >= safeTimeout {
                return nil
            }

            Thread.sleep(forTimeInterval: 0.01)
        }
    }

    func loadBuffer(_ data: Data) {
        lock.lock()
        defer { lock.unlock() }
        rxBytes = Array(data)
        rxCounter = 0
        rxTsMs = Array(repeating: 0, count: rxBytes.count / Self.packetSizeBytes)
    }

    func getBuffer() -> Data {
        lock.lock()
        defer { lock.unlock() }
        return Data(rxBytes)
    }

    func readRxSince(packetIndex: UInt64, maxPackets: Int) -> NativeBufferRust.ReadPackets {
        lock.lock()
        defer { lock.unlock() }
        return readSince(bytes: rxBytes, timestamps: rxTsMs, packetIndex: packetIndex, maxPackets: maxPackets)
    }

    func readTxSince(packetIndex: UInt64, maxPackets: Int) -> NativeBufferRust.ReadPackets {
        lock.lock()
        defer { lock.unlock() }
        return readSince(bytes: txBytes, timestamps: txTsMs, packetIndex: packetIndex, maxPackets: maxPackets)
    }

    func nextRxPacket() -> (packet64: Data, tsMs: UInt64)? {
        lock.lock()
        defer { lock.unlock() }

        let packets = UInt64(rxBytes.count / Self.packetSizeBytes)
        if rxCounter >= packets { return nil }

        let startByte = Int(rxCounter) * Self.packetSizeBytes
        if startByte + Self.packetSizeBytes > rxBytes.count { return nil }

        let pkt = Data(rxBytes[startByte..<(startByte + Self.packetSizeBytes)])
        let ts = Int(rxCounter) < rxTsMs.count ? rxTsMs[Int(rxCounter)] : 0
        rxCounter += 1
        return (pkt, ts)
    }

    func compressDataBits(rangeStart: Int, rangeEnd: Int, numberBins: Int) -> ([Float], [Float]) {
        lock.lock()
        defer { lock.unlock() }

        let buffer = rxBytes
        let totalBits = buffer.count * 8
        if buffer.isEmpty || rangeStart >= rangeEnd || rangeStart >= totalBits || numberBins <= 0 {
            return ([], [])
        }

        let end = min(rangeEnd, totalBits)
        let start = min(rangeStart, end)
        let span = end - start
        if span <= 0 { return ([], []) }

        func bitAt(_ i: Int) -> UInt8 {
            let byteIndex = i >> 3
            if byteIndex < 0 || byteIndex >= buffer.count { return 0 }
            return (buffer[byteIndex] >> (i & 7)) & 1
        }

        var timeValues: [Float] = []
        var dataValues: [Float] = []

        if span <= numberBins * 2 {
            timeValues.reserveCapacity(span)
            dataValues.reserveCapacity(span)
            for i in start..<end {
                timeValues.append(Float(i))
                dataValues.append(bitAt(i) == 1 ? 255.0 : 0.0)
            }
            return (timeValues, dataValues)
        }

        let binWidth = Float(span) / Float(numberBins)
        for bin in 0..<numberBins {
            let binStart = Int(floor(Float(start) + Float(bin) * binWidth))
            var binEnd = Int(floor(Float(binStart) + binWidth))
            if binEnd > end { binEnd = end }
            if binEnd <= binStart { continue }

            var hasLow = false
            var hasHigh = false
            var i = binStart
            while i < binEnd {
                let byteIndex = i >> 3
                if byteIndex >= buffer.count { break }

                if (i & 7) == 0 && i + 8 <= binEnd {
                    let byteVal = buffer[byteIndex]
                    if byteVal == 0 {
                        hasLow = true
                    } else if byteVal == 255 {
                        hasHigh = true
                    } else {
                        hasLow = true
                        hasHigh = true
                    }
                    i += 8
                } else {
                    if bitAt(i) == 1 { hasHigh = true } else { hasLow = true }
                    i += 1
                }

                if hasLow && hasHigh { break }
            }

            if hasLow || hasHigh {
                timeValues.append(Float(binStart))
                dataValues.append(hasLow ? 0.0 : 255.0)
                timeValues.append(Float(binEnd - 1))
                dataValues.append(hasHigh ? 255.0 : 0.0)
            }
        }

        return (timeValues, dataValues)
    }

    func takeRxState() -> (rxBytes: Data, rxTsMs: [UInt64], rxCounter: UInt64) {
        lock.lock()
        defer { lock.unlock() }
        return (Data(rxBytes), rxTsMs, rxCounter)
    }

    func restoreRxState(rxBytes: Data, rxTsMs: [UInt64], rxCounter: UInt64) {
        lock.lock()
        defer { lock.unlock() }
        self.rxBytes = Array(rxBytes)
        self.rxTsMs = rxTsMs

        let packets = self.rxBytes.count / Self.packetSizeBytes
        if self.rxTsMs.count < packets {
            self.rxTsMs.append(contentsOf: Array(repeating: 0, count: packets - self.rxTsMs.count))
        } else if self.rxTsMs.count > packets {
            self.rxTsMs = Array(self.rxTsMs.prefix(packets))
        }

        self.rxCounter = min(rxCounter, UInt64(packets))
    }

    private func readSince(bytes: [UInt8], timestamps: [UInt64], packetIndex: UInt64, maxPackets: Int) -> NativeBufferRust.ReadPackets {
        let availablePackets = UInt64(bytes.count / Self.packetSizeBytes)
        if availablePackets == 0 || maxPackets <= 0 || packetIndex >= availablePackets {
            return NativeBufferRust.ReadPackets(data: [], ts_ms: [], next_packet_index: min(packetIndex, availablePackets), available_packets: availablePackets)
        }

        let toRead = UInt64(min(maxPackets, Int(availablePackets - packetIndex)))
        let startByte = Int(packetIndex) * Self.packetSizeBytes
        let endByte = min(bytes.count, startByte + Int(toRead) * Self.packetSizeBytes)
        let dataSlice = startByte < endByte ? Array(bytes[startByte..<endByte]) : []

        let tsStart = Int(packetIndex)
        let tsEnd = min(timestamps.count, tsStart + Int(toRead))
        let tsSlice = tsStart < tsEnd ? Array(timestamps[tsStart..<tsEnd]) : []

        return NativeBufferRust.ReadPackets(data: dataSlice, ts_ms: tsSlice, next_packet_index: packetIndex + toRead, available_packets: availablePackets)
    }
}
